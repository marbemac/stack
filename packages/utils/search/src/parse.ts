import type { ILexingResult, IRecognitionException, Lexer } from 'chevrotain';

import { SearchLexer } from './grammar/lexer.ts';
import { SearchParser } from './grammar/parser.ts';
import type { SearchString } from './types.ts';

let lexerSingleton: Lexer;
let parserSingleton: SearchParser;

export interface ParseQueryOpts<T extends InputType> {
  input: SearchString;
  inputType: T;
  lexer?: Lexer;
  parser?: SearchParser;
}

type InputType = 'searchQuery' | 'selectClause' | 'selectExpression' | 'fromClause' | 'whereClause' | 'whereExpression';

export const parseSearch = <T extends InputType>({
  input,
  inputType,
  ...rest
}: ParseQueryOpts<T>): {
  cst: ReturnType<SearchParser[T]>;
  errors: IRecognitionException[];
  lexResult: ILexingResult;
} => {
  let lexer = rest.lexer || lexerSingleton;
  if (!lexer) {
    lexer = lexerSingleton = new SearchLexer();
  }

  const lexResult = lexer.tokenize(input);

  let parser = rest.parser || parserSingleton;
  if (!parser) {
    parser = parserSingleton = new SearchParser();
  }

  // "input" is a setter which will reset the parser's state.
  parser.input = lexResult.tokens;

  const cst = parser[inputType]() as ReturnType<SearchParser[T]>;

  return { cst, errors: parser.errors, lexResult };
};
