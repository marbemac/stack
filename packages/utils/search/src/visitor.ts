import type { CstNode } from 'chevrotain';

import { SearchParser } from './grammar/parser.ts';
import type {
  AtomicQualifierValCstChildren,
  AtomicQualifierValCstNode,
  FromClauseCstChildren,
  FromClauseCstNode,
  FunctionArgCstChildren,
  FunctionArgCstNode,
  FunctionCstChildren,
  FunctionCstNode,
  QualifierCstChildren,
  QualifierCstNode,
  QualifierInCstChildren,
  QualifierInCstNode,
  QualifierKeyCstChildren,
  QualifierKeyCstNode,
  QualifierOpCstChildren,
  QualifierOpCstNode,
  QualifierValCstChildren,
  QualifierValCstNode,
  RelativeDateValCstChildren,
  RelativeDateValCstNode,
  SearchQueryCstChildren,
  SearchQueryCstNode,
  SelectClauseCstChildren,
  SelectClauseCstNode,
  SelectExpressionCstChildren,
  TSearchCstVisitor,
  WhereClauseCstChildren,
  WhereClauseCstNode,
  WhereExpressionCstChildren,
  WhereExpressionCstNode,
} from './grammar/parser-cst-types.ts';

export const enum SearchToken {
  SearchQuery = 'searchQuery',
  SelectClause = 'selectClause',
  FromClause = 'fromClause',
  WhereClause = 'whereClause',
  Qualifier = 'qualifier',
  Function = 'function',
  QualifierKey = 'qualifierKey',
  QualifierIn = 'qualifierIn',
  RelativeDateVal = 'relativeDateVal',
  TextVal = 'textVal',
  NumberVal = 'numberVal',
}

/**
 * Object representing an invalid token state.
 *
 * Provide custom types or expectedTypes via the generics.
 */
export interface InvalidToken<T = unknown, ET = unknown> {
  /**
   * The message indicating why the token is invalid
   */
  reason: string;

  /**
   * The invalid reason type
   */
  type: T;

  /**
   * In the case where a filter is invalid, we may be expecting a different
   * type for this filter based on the key. This can be useful to hint to the
   * user what values they should be providing.
   *
   * This may be multiple filter types.
   */
  expectedType?: ET[];
}

interface SearchTokenBase<T = unknown, ET = unknown> {
  invalid?: InvalidToken<T, ET>;
}

export type AnySearchToken<T = unknown, ET = unknown> =
  | SearchQueryAstNode<T, ET>
  | SelectClauseAstNode<T, ET>
  | FromClauseAstNode<T, ET>
  | WhereClauseAstNode<T, ET>
  | QualifierAstNode<T, ET>
  | FunctionAstNode<T, ET>
  | QualifierKeyAstNode<T, ET>
  | QualifierInAstNode<T, ET>
  | RelativeDateValAstNode<T, ET>
  | AtomicQualifierValAstNode<T, ET>;

export interface SearchQueryAstNode<T = unknown, ET = unknown> extends SearchTokenBase<T, ET> {
  type: SearchToken.SearchQuery;
  selectClause: SelectClauseAstNode<T, ET>;
  fromClause: FromClauseAstNode<T, ET>;
  whereClause: WhereClauseAstNode<T, ET>;
}

export interface SelectClauseAstNode<T = unknown, ET = unknown> extends SearchTokenBase<T, ET> {
  type: SearchToken.SelectClause;
  columns: SelectExpressionAstColumns<T, ET>;
}

export type SelectExpressionAstColumns<T = unknown, ET = unknown> = (
  | QualifierAstNode<T, ET>
  | FunctionAstNode<T, ET>
  | AtomicQualifierValAstNode<T, ET>
)[];

export interface FromClauseAstNode<T = unknown, ET = unknown> extends SearchTokenBase<T, ET> {
  type: SearchToken.FromClause;
  table: string;
}

export interface WhereClauseAstNode<T = unknown, ET = unknown> extends SearchTokenBase<T, ET> {
  type: SearchToken.WhereClause;
  conditions: WhereExpressionAstConditions<T, ET>;
}

export type WhereExpressionAstConditions<T = unknown, ET = unknown> = (
  | QualifierAstNode<T, ET>
  | FunctionAstNode<T, ET>
  | AtomicQualifierValAstNode<T, ET>
)[];

export interface QualifierAstNode<T = unknown, ET = unknown> extends SearchTokenBase<T, ET> {
  type: SearchToken.Qualifier;
  negated: boolean;
  lhs: QualifierKeyAstNode<T, ET>;
  op?: QualifierOp;
  rhs?: QualifierRhs<T, ET>;
}

export interface FunctionAstNode<T = unknown, ET = unknown> extends SearchTokenBase<T, ET> {
  type: SearchToken.Function;
  name: string;
  negated: boolean;
  args: FunctionAstArgs<T, ET>[];
  op?: QualifierOp;
  rhs?: QualifierRhs<T, ET>;
}

export type FunctionAstArgs<T = unknown, ET = unknown> = (QualifierAstNode<T, ET> | AtomicQualifierValAstNode<T, ET>)[];

export interface QualifierKeyAstNode<T = unknown, ET = unknown> extends SearchTokenBase<T, ET> {
  type: SearchToken.QualifierKey;
  value: string;
}

export type QualifierRhs<T = unknown, ET = unknown> =
  | QualifierInAstNode<T, ET>
  | AtomicQualifierValAstNode<T, ET>
  | RelativeDateValAstNode<T, ET>;

export interface QualifierInAstNode<T = unknown, ET = unknown> extends SearchTokenBase<T, ET> {
  type: SearchToken.QualifierIn;
  values: AtomicQualifierValAstNode<T, ET>[];
}

export interface RelativeDateValAstNode<T = unknown, ET = unknown> extends SearchTokenBase<T, ET> {
  type: SearchToken.RelativeDateVal;
  value: string;
  sign: '+' | '-';
  unit: 's' | 'mi' | 'h' | 'd' | 'w' | 'm' | 'q' | 'y';
}

export interface TextValAstNode<T = unknown, ET = unknown> extends SearchTokenBase<T, ET> {
  type: SearchToken.TextVal;
  quoted?: boolean;
  value: string;
}

export interface NumberValAstNode<T = unknown, ET = unknown> extends SearchTokenBase<T, ET> {
  type: SearchToken.TextVal;
  value: string;
}

export interface AtomicQualifierValAstNode<T = unknown, ET = unknown> extends SearchTokenBase<T, ET> {
  type: SearchToken.TextVal | SearchToken.NumberVal;
  quoted?: boolean;
  value: string;
}

export type QualifierOp = '=' | '>' | '<' | '>=' | '<=';

export type SearchVisitor = ReturnType<typeof createSearchVisitor>;

const checkNodeType = (node: unknown, type: string): boolean =>
  !!node && typeof node === 'object' && 'type' in node && node.type === type;

export const isQualifierNode = (node: unknown): node is QualifierAstNode => checkNodeType(node, SearchToken.Qualifier);

export const isFunctionNode = (node: unknown): node is FunctionAstNode => checkNodeType(node, SearchToken.Function);

export const isQualifierKeyNode = (node: unknown): node is QualifierKeyAstNode =>
  checkNodeType(node, SearchToken.QualifierKey);

export const isQualifierInNode = (node: unknown): node is QualifierInAstNode =>
  checkNodeType(node, SearchToken.QualifierIn);

export const isAtomicQualifierValNode = (node: unknown): node is AtomicQualifierValAstNode =>
  isTextValNode(node) || isNumberValNode(node);

export const isTextValNode = (node: unknown): node is TextValAstNode => checkNodeType(node, SearchToken.TextVal);

export const isNumberValNode = (node: unknown): node is NumberValAstNode => checkNodeType(node, SearchToken.NumberVal);

let parserSingleton: SearchParser;

interface CreateSearchVisitorOpts {
  transform?: <T extends AnySearchToken>(node: T) => T;
}

export const createSearchVisitor = ({ transform }: CreateSearchVisitorOpts = {}) => {
  function maybeTransform<T extends AnySearchToken>(node: T): T {
    return transform ? transform(node) : node;
  }

  if (!parserSingleton) {
    parserSingleton = new SearchParser();
  }

  const BaseVisitor = parserSingleton.getBaseCstVisitorConstructor();

  /** This simply helps keep this visitor typings tight. Maps CST nodes to AST nodes. */
  type GetReturnType<T> = T extends SearchQueryCstNode
    ? SearchQueryAstNode
    : T extends SelectClauseCstNode | SelectClauseCstNode[]
      ? SelectClauseAstNode
      : T extends FromClauseCstNode | FromClauseCstNode[]
        ? FromClauseAstNode
        : T extends WhereClauseCstNode | WhereClauseCstNode[]
          ? WhereClauseAstNode
          : T extends WhereExpressionCstNode | WhereExpressionCstNode[]
            ? WhereExpressionAstConditions
            : T extends QualifierCstNode | QualifierCstNode[]
              ? QualifierAstNode
              : T extends FunctionCstNode | FunctionCstNode[]
                ? FunctionAstNode
                : T extends FunctionArgCstNode | FunctionArgCstNode[]
                  ? FunctionAstArgs
                  : T extends QualifierKeyCstNode | QualifierKeyCstNode[]
                    ? QualifierKeyAstNode
                    : T extends QualifierValCstNode | QualifierValCstNode[]
                      ? QualifierRhs
                      : T extends QualifierInCstNode | QualifierInCstNode[]
                        ? QualifierInAstNode
                        : T extends RelativeDateValCstNode | RelativeDateValCstNode[]
                          ? RelativeDateValAstNode
                          : T extends AtomicQualifierValCstNode | AtomicQualifierValCstNode[]
                            ? AtomicQualifierValAstNode
                            : T extends QualifierOpCstNode | QualifierOpCstNode[]
                              ? QualifierOp
                              : never;

  type VisitFn = <T extends CstNode | CstNode[]>(cstNode: T, param?: unknown) => GetReturnType<T>;

  class SearchCstVisitor extends BaseVisitor implements TSearchCstVisitor<unknown, unknown> {
    #visit: VisitFn;

    constructor() {
      super();
      this.#visit = this.visit as VisitFn;
      this.validateVisitor();
    }

    searchQuery(ctx: SearchQueryCstChildren) {
      return maybeTransform({
        type: SearchToken.SearchQuery as const,
        selectClause: this.#visit(ctx.selectClause),
        fromClause: this.#visit(ctx.fromClause),
        whereClause: this.#visit(ctx.whereClause!),
      }) satisfies SearchQueryAstNode;
    }

    selectClause(ctx: SelectClauseCstChildren) {
      return maybeTransform({
        type: SearchToken.SelectClause as const,
        columns: this.#visit(ctx.selectExpression),
      }) satisfies SelectClauseAstNode;
    }

    selectExpression(ctx: SelectExpressionCstChildren) {
      return (ctx.columns || []).map(f => this.#visit(f)) satisfies SelectExpressionAstColumns;
    }

    fromClause(ctx: FromClauseCstChildren) {
      return maybeTransform({
        type: SearchToken.FromClause as const,
        table: ctx.Identifier[0]?.image || '',
      }) satisfies FromClauseAstNode;
    }

    whereClause(ctx: WhereClauseCstChildren) {
      return maybeTransform({
        type: SearchToken.WhereClause as const,
        conditions: this.#visit(ctx.whereExpression),
      }) satisfies WhereClauseAstNode;
    }

    whereExpression(ctx: WhereExpressionCstChildren) {
      return (ctx.conditions || []).map(f => this.#visit(f)) satisfies WhereExpressionAstConditions;
    }

    qualifier(ctx: QualifierCstChildren) {
      return maybeTransform({
        type: SearchToken.Qualifier as const,
        negated: !!ctx.Negate,
        lhs: this.#visit(ctx.lhs!),
        op: ctx.qualifierOp ? this.#visit(ctx.qualifierOp) : undefined,
        rhs: ctx.rhs ? this.#visit(ctx.rhs) : undefined,
      }) satisfies QualifierAstNode;
    }

    function(ctx: FunctionCstChildren) {
      return maybeTransform({
        type: SearchToken.Function as const,
        name: ctx.Identifier[0]!.image,
        negated: !!ctx.Negate,
        args: (ctx.functionArg || []).map(arg => this.#visit(arg)).filter(arg => !!arg.length),
        op: ctx.qualifierOp ? this.#visit(ctx.qualifierOp) : undefined,
        rhs: ctx.rhs ? this.#visit(ctx.rhs) : undefined,
      }) satisfies FunctionAstNode;
    }

    functionArg(ctx: FunctionArgCstChildren) {
      return (ctx.args || []).map(arg => this.#visit(arg)) satisfies FunctionAstArgs;
    }

    qualifierKey(ctx: QualifierKeyCstChildren) {
      return maybeTransform({
        type: SearchToken.QualifierKey as const,
        value: (ctx.QualifierKey?.[0]?.image || '').slice(0, -1), // slice the ":" off
      }) satisfies QualifierKeyAstNode;
    }

    qualifierVal(ctx: QualifierValCstChildren) {
      if (!ctx.val) return undefined;

      return maybeTransform(this.#visit(ctx.val)) satisfies QualifierRhs;
    }

    qualifierIn(ctx: QualifierInCstChildren) {
      return maybeTransform({
        type: SearchToken.QualifierIn as const,
        values: ctx.atomicQualifierVal.map(val => this.#visit(val)),
      }) satisfies QualifierInAstNode;
    }

    relativeDateVal(ctx: RelativeDateValCstChildren) {
      return maybeTransform({
        type: SearchToken.RelativeDateVal as const,
        value: ctx.Number?.[0]?.image || '',
        sign: ctx.op?.[0]?.image as RelativeDateValAstNode['sign'],
        unit: ctx.DateUnit?.[0]?.image as RelativeDateValAstNode['unit'],
      }) satisfies RelativeDateValAstNode;
    }

    atomicQualifierVal(ctx: AtomicQualifierValCstChildren) {
      const type = ctx.Number ? SearchToken.NumberVal : SearchToken.TextVal;
      let target = ctx.Identifier?.[0] || ctx.Number?.[0];

      let quoted = false;
      if (ctx.QuotedIdentifier) {
        quoted = true;
        target = ctx.QuotedIdentifier?.[0];
      }

      return maybeTransform({
        type,
        quoted,
        value: target?.image || '',
      }) satisfies AtomicQualifierValAstNode;
    }

    qualifierOp(ctx: QualifierOpCstChildren) {
      return (ctx.op?.[0]?.image || '') as QualifierOp;
    }
  }

  return new SearchCstVisitor();
};
