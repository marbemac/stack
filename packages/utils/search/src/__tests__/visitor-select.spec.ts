import { describe, expect, it } from 'vitest';

import { parseSearch } from '../parse.ts';
import { createSearchVisitor } from '../visitor.ts';
import { caseGroups } from './fixtures/common.ts';

const prettyPrintParseResult = (input: string) => {
  const { cst, errors } = parseSearch({ input: input, inputType: 'selectExpr' });

  const visitor = createSearchVisitor({
    /**
     * Just cleans up the snapshots a little bit for easier review
     */
    transform: node => {
      // @ts-expect-error remove `quoted` property if false, since that's the default
      if (node.quoted === false) delete node.quoted;

      // @ts-expect-error remove `negated` property if false, since that's the default
      if (node.negated === false) delete node.negated;

      // @ts-expect-error remove `sort` property if undefined, since that's the default
      if (typeof node.sort === 'undefined') delete node.sort;

      if (node.isBranchInvalid === false) {
        // @ts-expect-error remove `isBranchInvalid` property if false, since that's the default
        delete node.isBranchInvalid;
      }

      return node;
    },
  });

  const ast = visitor.selectExpr(cst.children);

  return { errors, ast };
};

describe.each(Object.entries(caseGroups))('select %s', (_, cases) => {
  it.each(cases)('%s', (_, input) => {
    expect({
      input,
      result: prettyPrintParseResult(input),
    }).toMatchSnapshot();
  });
});

describe('sorting', () => {
  it('supports text vals', () => {
    expect(prettyPrintParseResult('+name').ast.columns[0]).toMatchObject({ sort: 'asc' });
  });

  it('supports desc', () => {
    expect(prettyPrintParseResult('-name').ast.columns[0]).toMatchObject({ sort: 'desc' });
  });

  it('supports functions', () => {
    expect(prettyPrintParseResult('+count()').ast.columns[0]).toMatchObject({ sort: 'asc' });
  });
});

it('supports tags that start with a $', () => {
  expect(prettyPrintParseResult('$first_seen').ast.columns[0]).toMatchObject({
    value: '$first_seen',
  });
});
