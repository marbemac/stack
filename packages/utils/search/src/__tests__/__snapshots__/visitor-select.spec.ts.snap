// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`select freeform > mixed throughout 1`] = `
{
  "input": "freeform1 age:5 freeform2 search",
  "result": {
    "ast": [
      {
        "type": "textVal",
        "value": "freeform1",
      },
      {
        "lhs": {
          "type": "qualifierKey",
          "value": "age",
        },
        "op": "",
        "rhs": {
          "type": "numberVal",
          "value": "5",
        },
        "type": "qualifier",
      },
      {
        "type": "textVal",
        "value": "freeform2",
      },
      {
        "type": "textVal",
        "value": "search",
      },
    ],
    "errors": [],
  },
}
`;

exports[`select freeform > quoted 1`] = `
{
  "input": ""a quoted:freeform" age:5",
  "result": {
    "ast": [
      {
        "quoted": true,
        "type": "textVal",
        "value": "a quoted:freeform",
      },
      {
        "lhs": {
          "type": "qualifierKey",
          "value": "age",
        },
        "op": "",
        "rhs": {
          "type": "numberVal",
          "value": "5",
        },
        "type": "qualifier",
      },
    ],
    "errors": [],
  },
}
`;

exports[`select freeform > treats as a string 1`] = `
{
  "input": "freeform",
  "result": {
    "ast": [
      {
        "type": "textVal",
        "value": "freeform",
      },
    ],
    "errors": [],
  },
}
`;

exports[`select functions > no args 1`] = `
{
  "input": "full_name()",
  "result": {
    "ast": [
      {
        "args": [],
        "name": "full_name",
        "op": undefined,
        "rhs": undefined,
        "type": "function",
      },
    ],
    "errors": [],
  },
}
`;

exports[`select functions > no args with space 1`] = `
{
  "input": "full_name(  )",
  "result": {
    "ast": [
      {
        "args": [],
        "name": "full_name",
        "op": undefined,
        "rhs": undefined,
        "type": "function",
      },
    ],
    "errors": [],
  },
}
`;

exports[`select functions > relatvie date rhs 1`] = `
{
  "input": "last_seen():-7d",
  "result": {
    "ast": [
      {
        "args": [],
        "name": "last_seen",
        "op": "",
        "rhs": {
          "sign": "-",
          "type": "relativeDateVal",
          "unit": "d",
          "value": "7",
        },
        "type": "function",
      },
    ],
    "errors": [],
  },
}
`;

exports[`select functions > single string arg 1`] = `
{
  "input": "count(events)",
  "result": {
    "ast": [
      {
        "args": [
          [
            {
              "type": "textVal",
              "value": "events",
            },
          ],
        ],
        "name": "count",
        "op": undefined,
        "rhs": undefined,
        "type": "function",
      },
    ],
    "errors": [],
  },
}
`;

exports[`select functions > with filter argument 1`] = `
{
  "input": "count(projects, plan:free !has:email):>5",
  "result": {
    "ast": [
      {
        "args": [
          [
            {
              "type": "textVal",
              "value": "projects",
            },
          ],
          [
            {
              "lhs": {
                "type": "qualifierKey",
                "value": "plan",
              },
              "op": "",
              "rhs": {
                "type": "textVal",
                "value": "free",
              },
              "type": "qualifier",
            },
            {
              "lhs": {
                "type": "qualifierKey",
                "value": "has",
              },
              "negated": true,
              "op": "",
              "rhs": {
                "type": "textVal",
                "value": "email",
              },
              "type": "qualifier",
            },
          ],
        ],
        "name": "count",
        "op": ">",
        "rhs": {
          "type": "numberVal",
          "value": "5",
        },
        "type": "function",
      },
    ],
    "errors": [],
  },
}
`;

exports[`select functions > with quoted argument 1`] = `
{
  "input": "count(projects, "free search"):>0.5",
  "result": {
    "ast": [
      {
        "args": [
          [
            {
              "type": "textVal",
              "value": "projects",
            },
          ],
          [
            {
              "quoted": true,
              "type": "textVal",
              "value": "free search",
            },
          ],
        ],
        "name": "count",
        "op": ">",
        "rhs": {
          "type": "numberVal",
          "value": "0.5",
        },
        "type": "function",
      },
    ],
    "errors": [],
  },
}
`;

exports[`select functions > with rhs 1`] = `
{
  "input": "divide(projects, max_projects):>0.5",
  "result": {
    "ast": [
      {
        "args": [
          [
            {
              "type": "textVal",
              "value": "projects",
            },
          ],
          [
            {
              "type": "textVal",
              "value": "max_projects",
            },
          ],
        ],
        "name": "divide",
        "op": ">",
        "rhs": {
          "type": "numberVal",
          "value": "0.5",
        },
        "type": "function",
      },
    ],
    "errors": [],
  },
}
`;

exports[`select in filters > negation 1`] = `
{
  "input": "!user:[jane, john]",
  "result": {
    "ast": [
      {
        "lhs": {
          "type": "qualifierKey",
          "value": "user",
        },
        "negated": true,
        "op": "",
        "rhs": {
          "type": "qualifierIn",
          "values": [
            {
              "type": "textVal",
              "value": "jane",
            },
            {
              "type": "textVal",
              "value": "john",
            },
          ],
        },
        "type": "qualifier",
      },
    ],
    "errors": [],
  },
}
`;

exports[`select in filters > numbers 1`] = `
{
  "input": "release:[12.0]",
  "result": {
    "ast": [
      {
        "lhs": {
          "type": "qualifierKey",
          "value": "release",
        },
        "op": "",
        "rhs": {
          "type": "qualifierIn",
          "values": [
            {
              "type": "numberVal",
              "value": "12.0",
            },
          ],
        },
        "type": "qualifier",
      },
    ],
    "errors": [],
  },
}
`;

exports[`select in filters > quoted 1`] = `
{
  "input": "release:[jane, "john doe"]",
  "result": {
    "ast": [
      {
        "lhs": {
          "type": "qualifierKey",
          "value": "release",
        },
        "op": "",
        "rhs": {
          "type": "qualifierIn",
          "values": [
            {
              "type": "textVal",
              "value": "jane",
            },
            {
              "quoted": true,
              "type": "textVal",
              "value": "john doe",
            },
          ],
        },
        "type": "qualifier",
      },
    ],
    "errors": [],
  },
}
`;

exports[`select in filters > works 1`] = `
{
  "input": "user:[jane, john]",
  "result": {
    "ast": [
      {
        "lhs": {
          "type": "qualifierKey",
          "value": "user",
        },
        "op": "",
        "rhs": {
          "type": "qualifierIn",
          "values": [
            {
              "type": "textVal",
              "value": "jane",
            },
            {
              "type": "textVal",
              "value": "john",
            },
          ],
        },
        "type": "qualifier",
      },
    ],
    "errors": [],
  },
}
`;

exports[`select operators > < 1`] = `
{
  "input": "num_members:<5.0",
  "result": {
    "ast": [
      {
        "lhs": {
          "type": "qualifierKey",
          "value": "num_members",
        },
        "op": "<",
        "rhs": {
          "type": "numberVal",
          "value": "5.0",
        },
        "type": "qualifier",
      },
    ],
    "errors": [],
  },
}
`;

exports[`select operators > <= 1`] = `
{
  "input": "num_members:<=5.0",
  "result": {
    "ast": [
      {
        "lhs": {
          "type": "qualifierKey",
          "value": "num_members",
        },
        "op": "<=",
        "rhs": {
          "type": "numberVal",
          "value": "5.0",
        },
        "type": "qualifier",
      },
    ],
    "errors": [],
  },
}
`;

exports[`select operators > = 1`] = `
{
  "input": "num_members:=5",
  "result": {
    "ast": [
      {
        "lhs": {
          "type": "qualifierKey",
          "value": "num_members",
        },
        "op": "=",
        "rhs": {
          "type": "numberVal",
          "value": "5",
        },
        "type": "qualifier",
      },
    ],
    "errors": [],
  },
}
`;

exports[`select operators > > 1`] = `
{
  "input": "num_members:>5.0",
  "result": {
    "ast": [
      {
        "lhs": {
          "type": "qualifierKey",
          "value": "num_members",
        },
        "op": ">",
        "rhs": {
          "type": "numberVal",
          "value": "5.0",
        },
        "type": "qualifier",
      },
    ],
    "errors": [],
  },
}
`;

exports[`select operators > >= 1`] = `
{
  "input": "num_members:>=5",
  "result": {
    "ast": [
      {
        "lhs": {
          "type": "qualifierKey",
          "value": "num_members",
        },
        "op": ">=",
        "rhs": {
          "type": "numberVal",
          "value": "5",
        },
        "type": "qualifier",
      },
    ],
    "errors": [],
  },
}
`;

exports[`select relative date > + 1`] = `
{
  "input": "renews_at:+30d",
  "result": {
    "ast": [
      {
        "lhs": {
          "type": "qualifierKey",
          "value": "renews_at",
        },
        "op": "",
        "rhs": {
          "sign": "+",
          "type": "relativeDateVal",
          "unit": "d",
          "value": "30",
        },
        "type": "qualifier",
      },
    ],
    "errors": [],
  },
}
`;

exports[`select relative date > - 1`] = `
{
  "input": "last_seen:-1d",
  "result": {
    "ast": [
      {
        "lhs": {
          "type": "qualifierKey",
          "value": "last_seen",
        },
        "op": "",
        "rhs": {
          "sign": "-",
          "type": "relativeDateVal",
          "unit": "d",
          "value": "1",
        },
        "type": "qualifier",
      },
    ],
    "errors": [],
  },
}
`;

exports[`select relative date > decimal 1`] = `
{
  "input": "renews_at:+1.2s",
  "result": {
    "ast": [
      {
        "lhs": {
          "type": "qualifierKey",
          "value": "renews_at",
        },
        "op": "",
        "rhs": {
          "sign": "+",
          "type": "relativeDateVal",
          "unit": "s",
          "value": "1.2",
        },
        "type": "qualifier",
      },
    ],
    "errors": [],
  },
}
`;

exports[`select relative date > with other filters 1`] = `
{
  "input": "plan:free renews_at:+4w "free form"",
  "result": {
    "ast": [
      {
        "lhs": {
          "type": "qualifierKey",
          "value": "plan",
        },
        "op": "",
        "rhs": {
          "type": "textVal",
          "value": "free",
        },
        "type": "qualifier",
      },
      {
        "lhs": {
          "type": "qualifierKey",
          "value": "renews_at",
        },
        "op": "",
        "rhs": {
          "sign": "+",
          "type": "relativeDateVal",
          "unit": "w",
          "value": "4",
        },
        "type": "qualifier",
      },
      {
        "quoted": true,
        "type": "textVal",
        "value": "free form",
      },
    ],
    "errors": [],
  },
}
`;

exports[`select text filters > .-_[0-9] in keys 1`] = `
{
  "input": "payments.plan-type_is2:free",
  "result": {
    "ast": [
      {
        "lhs": {
          "type": "qualifierKey",
          "value": "payments.plan-type_is2",
        },
        "op": "",
        "rhs": {
          "type": "textVal",
          "value": "free",
        },
        "type": "qualifier",
      },
    ],
    "errors": [],
  },
}
`;

exports[`select text filters > @.-_ in values 1`] = `
{
  "input": "email:jane@example.com type:public-company_x",
  "result": {
    "ast": [
      {
        "lhs": {
          "type": "qualifierKey",
          "value": "email",
        },
        "op": "",
        "rhs": {
          "type": "textVal",
          "value": "jane@example.com",
        },
        "type": "qualifier",
      },
      {
        "lhs": {
          "type": "qualifierKey",
          "value": "type",
        },
        "op": "",
        "rhs": {
          "type": "textVal",
          "value": "public-company_x",
        },
        "type": "qualifier",
      },
    ],
    "errors": [],
  },
}
`;

exports[`select text filters > multiple 1`] = `
{
  "input": "plan:free age:5",
  "result": {
    "ast": [
      {
        "lhs": {
          "type": "qualifierKey",
          "value": "plan",
        },
        "op": "",
        "rhs": {
          "type": "textVal",
          "value": "free",
        },
        "type": "qualifier",
      },
      {
        "lhs": {
          "type": "qualifierKey",
          "value": "age",
        },
        "op": "",
        "rhs": {
          "type": "numberVal",
          "value": "5",
        },
        "type": "qualifier",
      },
    ],
    "errors": [],
  },
}
`;

exports[`select text filters > negation 1`] = `
{
  "input": "!plan:free",
  "result": {
    "ast": [
      {
        "lhs": {
          "type": "qualifierKey",
          "value": "plan",
        },
        "negated": true,
        "op": "",
        "rhs": {
          "type": "textVal",
          "value": "free",
        },
        "type": "qualifier",
      },
    ],
    "errors": [],
  },
}
`;

exports[`select text filters > negation in quoted 1`] = `
{
  "input": "!user:"john doe"",
  "result": {
    "ast": [
      {
        "lhs": {
          "type": "qualifierKey",
          "value": "user",
        },
        "negated": true,
        "op": "",
        "rhs": {
          "quoted": true,
          "type": "textVal",
          "value": "john doe",
        },
        "type": "qualifier",
      },
    ],
    "errors": [],
  },
}
`;

exports[`select text filters > partial filter still parses 1`] = `
{
  "input": "age:> plan:free user:",
  "result": {
    "ast": [
      {
        "lhs": {
          "type": "qualifierKey",
          "value": "age",
        },
        "op": ">",
        "rhs": undefined,
        "type": "qualifier",
      },
      {
        "lhs": {
          "type": "qualifierKey",
          "value": "plan",
        },
        "op": "",
        "rhs": {
          "type": "textVal",
          "value": "free",
        },
        "type": "qualifier",
      },
      {
        "lhs": {
          "type": "qualifierKey",
          "value": "user",
        },
        "op": undefined,
        "rhs": undefined,
        "type": "qualifier",
      },
    ],
    "errors": [
      [NoViableAltException: Expecting: one of these possible Token sequences:
  1. [LBracket]
  2. [+]
  3. [Minus]
  4. [LQuote]
  5. [Identifier]
  6. [Number]
but found: 'plan:'],
    ],
  },
}
`;

exports[`select text filters > quoted value 1`] = `
{
  "input": "user:"john doe"",
  "result": {
    "ast": [
      {
        "lhs": {
          "type": "qualifierKey",
          "value": "user",
        },
        "op": "",
        "rhs": {
          "quoted": true,
          "type": "textVal",
          "value": "john doe",
        },
        "type": "qualifier",
      },
    ],
    "errors": [],
  },
}
`;

exports[`select text filters > works 1`] = `
{
  "input": "plan:free",
  "result": {
    "ast": [
      {
        "lhs": {
          "type": "qualifierKey",
          "value": "plan",
        },
        "op": "",
        "rhs": {
          "type": "textVal",
          "value": "free",
        },
        "type": "qualifier",
      },
    ],
    "errors": [],
  },
}
`;
